### PR TITLE
[feat]: [CDS-79379]: Stage execution entity to info for custom stage

### PR DESCRIPTION
**Pull request by lovish1999**
## Describe your changes
**Summary of the changes:** 

We have moved the custom stage step from pipeline service to ng-manager service. 

The custom stage step in pipeline service is using StageExecutionEntity class to store info in mongodb and that in turn is used to update the timescale db collections. 
But we have StageExecutionInfo class in ng-manager which we will be now using for the custom stage step in ng-manager. 

The only field missing in StageExecutionInfo is stageType. So we have added this new field in StageExecutionInfo. 

Earlier `StageExecutionHandler` in change data capture was not designed to handle this new field for StageExecutionInfo. Now we have added the implementation to handle the stage type as Custom. 

Also, since the update method in StageExecutionInfoService is actually performing an upsert operation, we have renamed the method accordingly. 

`CDStageExecutionHandler` - we only want to use this handler when the stage type is not custom as for custom stage, we have a separate handler with similar functionality - `CustomStageExecutionHandler`

So we have added a check that if the stage type is custom , we will not use this handler for change updates in StageExecutionInfo.  

Also, earlier `CustomStageExecutionHandler` was limited to failure info and stage id. Also, it was only handling change events for StageExecutionEntity class. Now, we have modified it to handle StageExecutionInfo class also. And also, now it saves infra execution and rollback details also in timescale db. 

Similar to Custom stage step in pipeline service, we use separate thread for this dashboard service tasks (to avoid pipeline failures) and accordingly changes have been made in `NextGenConfiguration` & `NextGenModule`.

## Checklist
- [ ] I've documented the changes in the PR description.
- [ ] I've tested this change either in PR or local environment.
- [ ] If hashcheck failed I followed [the checklist](https://harness.atlassian.net/wiki/spaces/DEL/pages/21016838831/PR+Codebasehash+Check+merge+checklist) and updated this PR description with my answers to make sure I'm not introducing any breaking changes.

## Comment Triggers
<details>
  <summary>Build triggers</summary>

- Feature build: `trigger feature-build`
- 
  Specific builds
  - `trigger manager`
  - `trigger dms`
  - `trigger ng_manager`
  - `trigger cvng `
  - `trigger cmdlib`
  - `trigger template_svc`
  - `trigger events_fmwrk_monitor`
  - `trigger event_server`
  - `trigger change_data_capture`
  -  Trigger multiple builds together: For eg: `trigger dms, manager`
- Immutable delegate `trigger publish-delegate`
</details>

## [Latest PR Check Triggers](https://github.com/harness/harness-core/blob/develop/.github/pull_request_template.md)

<details>
  <summary>PR Check triggers</summary>
You can run multiple PR check triggers by comma separating them in a single comment. e.g. `trigger ut0, ut1`

- Compile: `trigger compile`
- CodeformatCheckstyle: `trigger checkstylecodeformat`
  - CodeFormat: `trigger codeformat`
  - Checkstyle: `trigger checkstyle`
- MessageMetadata: `trigger messagecheck`
- File-Permission-Check: `trigger checkpermission`
- Recency: `trigger recency`
- BuildNumberMetadata: `trigger buildnum`
- Trigger CommonChecks: `trigger commonchecks`
- PMD: `trigger pmd`
- Copyright Check: `trigger copyrightcheck`
- Feature Name Check: `trigger featurenamecheck`
- UnitTests-ALL: `trigger uts`
- UnitTests-0: `trigger ut0`
- UnitTests-1: `trigger ut1`
- UnitTests-2: `trigger ut2`
- UnitTests-3: `trigger ut3`
- UnitTests-4: `trigger ut4`
- UnitTests-5: `trigger ut5`
- UnitTests-6: `trigger ut6`
- UnitTests-7: `trigger ut7`
- UnitTests-8: `trigger ut8`
- UnitTests-9: `trigger ut9`
- CodeBaseHash: `trigger codebasehash`
- CodeFormatCheckstyle: `trigger checkstylecodeformat`
- SonarScan: `trigger ss`
- GitLeaks: `trigger gitleaks`
- Trigger all Checks: `trigger smartchecks`
- Go Build: `trigger gobuild`
- Validate_Reviews: `trigger review`
- Module Dependency Check: `trigger mdc`
</details>

## PR check failures and solutions
https://harness.atlassian.net/wiki/spaces/BT/pages/21106884744/PR+Checks+-+Failures+and+Solutions

## [Contributor license agreement](https://github.com/harness/harness-core/blob/develop/CONTRIBUTOR_LICENSE_AGREEMENT.md)

- - -
This change is [Reviewable](https://reviewable.io/reviews/harness/harness-core/53527)

### CODE REVIEW

**Comment by lovish1999** [comment]
trigger uts

**Review by abhinavmittal-harness — CHANGES_REQUESTED** [review]
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/53527)* status: Some of the checks are too old. 27 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/pipeline/steps/CustomStageStep.java` line 44 at r1](https://reviewable.io/reviews/harness/harness-core/53527#-NgIgXL05KqPhdQtA8Cm:-NgIgXL05KqPhdQtA8Cn:br73832) ([raw file](https://github.com/harness/harness-core/blob/6954fc120e1c409d1433352b16fe6cea54d13bc4/125-cd-nextgen/modules/common/core/src/main/java/io/harness/cdng/pipeline/steps/CustomStageStep.java#L44)):*
> ```Java
> public class CustomStageStep implements ChildExecutable<StageElementParameters> {
>   public static final StepType STEP_TYPE =
>       StepType.newBuilder().setType("CUSTOM_STAGE").setStepCategory(StepCategory.STAGE).build();
> ```

Please see if there is a constant that can be used.
Same can be done in CDC service as well

**Review by abhinavmittal-harness — COMMENTED** [review]
> Reviewed all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/53527)* status: Some of the checks are too old. 27 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`110-change-data-capture/src/main/java/io/harness/changehandlers/CDStageExecutionHandler.java` line 43 at r1](https://reviewable.io/reviews/harness/harness-core/53527#-NgIgw437ODqL14OUcl_:-NgIgw437ODqL14OUcla:bqmd5oc) ([raw file](https://github.com/harness/harness-core/blob/6954fc120e1c409d1433352b16fe6cea54d13bc4/110-change-data-capture/src/main/java/io/harness/changehandlers/CDStageExecutionHandler.java#L43)):*
> ```Java
>     // for custom stage, separate CustomStageExecutionHandler is being used
>     String stageType = dbObject.get(StageExecutionInfoKeys.stageType).toString();
>     if ((stageType != null) && ("CUSTOM_STAGE".equals(stageType.toString()))) {
> ```

nit: Would prefer if we use stage type == null and stageType = CD for better clarity

**Review by abhinavmittal-harness — CHANGES_REQUESTED** [review]
> Reviewed 2 of 27 files at r1.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/53527)* status: Some of the checks are too old. 27 file(s) to review. 1 change requests. 0 of 2 approvals obtained. 

___
*[`110-change-data-capture/src/main/java/io/harness/changehandlers/CustomStageExecutionHandler.java` line 43 at r1](https://reviewable.io/reviews/harness/harness-core/53527#-NgIhOST8W1VGRs1GRFp:-NgIhOST8W1VGRs1GRFq:b-qh70qu) ([raw file](https://github.com/harness/harness-core/blob/6954fc120e1c409d1433352b16fe6cea54d13bc4/110-change-data-capture/src/main/java/io/harness/changehandlers/CustomStageExecutionHandler.java#L43)):*
> ```Java
> 
>     String stageExecutionId = getStageExecutionId(changeEvent, dbObject);
>     columnValueMapping.put("id", stageExecutionId);
> ```

if stageExecutionId is null we should not proceed further
___
*[`110-change-data-capture/src/main/java/io/harness/changehandlers/CustomStageExecutionHandler.java` line 51 at r1](https://reviewable.io/reviews/harness/harness-core/53527#-NgIhkQaAkRyZTh5y195:-NgIhkQaAkRyZTh5y196:b1o66cx) ([raw file](https://github.com/harness/harness-core/blob/6954fc120e1c409d1433352b16fe6cea54d13bc4/110-change-data-capture/src/main/java/io/harness/changehandlers/CustomStageExecutionHandler.java#L51)):*
> ```Java
>       BasicDBObject executionSummaryDetails =
>           (BasicDBObject) dbObject.get(StageExecutionInfoKeys.executionSummaryDetails);
>       if (executionSummaryDetails == null) {
> ```

is it possible if no infra is used and no failure executionSummaryDetails could be null

**Comment by lovish1999** [comment]
trigger ut4

**Review by abhinavmittal-harness — COMMENTED** [review]
> Reviewed 11 of 27 files at r1.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/53527)* status: Some of the checks are too old. 27 file(s) to review. 1 change requests. 0 of 2 approvals obtained.

**Comment by lovish1999** [comment]
trigger messagecheck

File: 127-cd-nextgen-entities/src/main/java/io/harness/executions/steps/ExecutionNodeType.java [hunk 1]
```diff
@@ -176,7 +176,8 @@ public enum ExecutionNodeType {
   CUSTOM_STAGE_ENVIRONMENT("CUSTOM_STAGE_ENVIRONMENT", YamlTypes.ENVIRONMENT_YAML),
   ECS_SERVICE_SETUP("ECS_SERVICE_SETUP", YamlTypes.ECS_SERVICE_SETUP),
   ECS_UPGRADE_CONTAINER("ECS_UPGRADE_CONTAINER", YamlTypes.ECS_UPGRADE_CONTAINER),
-  ECS_BASIC_ROLLBACK("ECS_BASIC_ROLLBACK", YamlTypes.ECS_BASIC_ROLLBACK);
+  ECS_BASIC_ROLLBACK("ECS_BASIC_ROLLBACK", YamlTypes.ECS_BASIC_ROLLBACK),
+  CUSTOM_STAGE("CUSTOM_STAGE", "custom");
```
Comment (RishabhGupta34) [inline 127-cd-nextgen-entities/src/main/java/io/harness/executions/steps/ExecutionNodeType.java:180]: nit. can we create a yaml type and then use that value, like the general convention above being followed?

Comment (lovish1999) [reply 127-cd-nextgen-entities/src/main/java/io/harness/executions/steps/ExecutionNodeType.java:180]: It's similar to what is done for deploy stage - ```DEPLOYMENT_STAGE_STEP("DEPLOYMENT_STAGE_STEP", "deployment")```

**Review by abhinavmittal-harness — DISMISSED** [review]
> Reviewed 12 of 27 files at r1, 13 of 13 files at r2, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/harness/harness-core/53527)* status: 32 file(s) to review. 1. unresolved discussions. 1 of 2 approvals obtained. 

___
*[`110-change-data-capture/src/main/java/io/harness/changehandlers/CustomStageExecutionHandler.java` line 51 at r1](https://reviewable.io/reviews/harness/harness-core/53527#-NgIhkQaAkRyZTh5y195:-NgNI40Z5WtAEwfTkAkh:b-rnz9za) ([raw file](https://github.com/harness/harness-core/blob/6954fc120e1c409d1433352b16fe6cea54d13bc4/110-change-data-capture/src/main/java/io/harness/changehandlers/CustomStageExecutionHandler.java#L51)):*
<details><summary><i>Previously, lovish1999 (Lovish Bansal) wrote…</i></summary><blockquote>

in case of null information, we set it via ```ExecutionSummaryDetails executionSummaryDetails \= stageExecutionInfo.getExecutionSummaryDetails() == null  
? ExecutionSummaryDetails.builder().build(). So looks like this cannot be null. But it's a null check to prevent NPE. Similar to what is done in CDStageExecutionHandler.
</blockquote></details>

Just test once this case resolving for now

**Comment by lovish1999** [comment]
trigger uts

**Review by RishabhGupta34 — APPROVED** [review]

**Comment by lovish1999** [comment]
trigger uts

**Comment by lovish1999** [comment]
trigger ut1

**Comment by tchaurasiya** [comment]
LGTM

**Review by tchaurasiya — APPROVED** [review]

**Review by RishabhGupta34 — APPROVED** [review]

**Comment by lovish1999** [comment]
trigger review

**Pull request merged**